console/collections/src/kary_merkle_tree/mod.rs

```diff
@@ -0,0 +1,271 @@
+        for _ in 0..DEPTH {
+            // Compute the index of the sibling hash, if it exists.
+            if let Some(siblings) = siblings::<ARITY>(index) {
+                // Append the sibling hashes to the path.
+                let sibling_hashes = siblings.iter().map(|index| self.tree[*index]).collect::<Vec<_>>();
+
+                path.push(sibling_hashes);
+                // Compute the index of the parent hash, if it exists.
+                match parent::<ARITY>(index) {
+                    // Update the index to the parent index.
+                    Some(parent) => index = parent,
+                    // If the parent does not exist, the path is complete.
+                    None => break,
+                }
+            }
+        }
+
+        // If the Merkle path length is not equal to `DEPTH`, pad the path with the empty hash.
+        let empty_hashes = (0..ARITY.saturating_sub(1)).map(|_| self.empty_hash).collect::<Vec<_>>();
+        path.resize(DEPTH as usize, empty_hashes);
+
+        // Return the Merkle path.
+        KaryMerklePath::try_from((leaf_index as u64, path))
+    }
+
+    /// Returns `true` if the given Merkle path is valid for the given root and leaf.
+    pub fn verify(&self, path: &KaryMerklePath<PH, DEPTH, ARITY>, root: &PH::Hash, leaf: &LH::Leaf) -> bool {
+        path.verify(&self.leaf_hasher, &self.path_hasher, root, leaf)
+    }
+
+    /// Returns the Merkle root of the tree.
+    pub const fn root(&self) -> &PH::Hash {
+        &self.root
+    }
+
+    /// Returns the Merkle tree (excluding the hashes of the leaves).
+    pub fn tree(&self) -> &[PH::Hash] {
+        &self.tree
+    }
+
+    /// Returns the empty hash.
+    pub const fn empty_hash(&self) -> &PH::Hash {
+        &self.empty_hash
+    }
+
+    /// Returns the number of leaves in the Merkle tree.
+    pub const fn number_of_leaves(&self) -> usize {
+        self.number_of_leaves
+    }
+}
+
+/// Returns the depth of the tree, given the size of the tree.
+#[inline]
+#[allow(clippy::cast_possible_truncation)]
+fn tree_depth<const DEPTH: u8, const ARITY: u8>(tree_size: usize) -> Result<u8> {
+    let tree_size = u64::try_from(tree_size)?;
+
+    ensure!(tree_size < 4503599627370496_u64, "Tree size must be less than 2^52");
+
+    // Calculate the tree depth based on the tree size and arity.
+    // log_arity(tree_size) =  ln(tree_size) / ln(arity).
+    let tree_depth_float = (tree_size as f64).ln() / (ARITY as f64).ln();
+    let tree_depth = u8::try_from(tree_depth_float.floor() as u64)?;
+
+    // Ensure the tree depth is within the depth bound.
+    match tree_depth <= DEPTH {
+        // Return the tree depth.
+        true => Ok(tree_depth),
+        false => bail!("Merkle tree cannot exceed depth {DEPTH}: attempted to reach depth {tree_depth}"),
+    }
+}
+
+/// Returns the indexes of the children, given an index.
+fn child_indexes<const ARITY: u8>(index: usize) -> Vec<usize> {
```

Comment:
ideally this would return an iterator; shall I change this in a new commit or mark it as a TODO for later?

Reply:
Iterator is fine, go ahead